### PR TITLE
Improve CQ1

### PR DESCRIPTION
**Pull request by iamcarbon**
- Removes unused Guard methods
- Seals ManagedByteBuffer
- Updates System.Numerics.Vectors on .NETSTANDARD 1.1 to v4.5
- Add and use HashCodeHelpers overloads accepting 3 and 4 args 
- Tidy things up using modern language features
- Update XUnit and Moq

### CODE REVIEW

**Comment by iamcarbon** [comment]
Ready for review.

**Comment by iamcarbon** [comment]
Ready for review.

Ping @tocsoft @JimBobSquarePants

File: src/SixLabors.Core/MathF.cs [hunk 1]
```diff
@@ -3,7 +3,7 @@
 
 using System.Runtime.CompilerServices;
 
-#if NETCOREAPP2_0
+#if NETCOREAPP2_0 || NETCOREAPP2_1
```
Comment (JimBobSquarePants) [inline src/SixLabors.Core/MathF.cs]: I wonder if there's a smarter check we can do to avoid having `|| NETCOREAPP2_N+1`?

Comment (dlemstra) [reply src/SixLabors.Core/MathF.cs]: I use this in inside the csproj for Magick.NET:
```xml
  <PropertyGroup Condition="$(TargetFramework.StartsWith('netstandard'))">
    <DefineConstants>$(DefineConstants);NETSTANDARD</DefineConstants>
  </PropertyGroup>
```

I could create a PR for this gets merged or @iamcarbon could add it to the current PR if he prefers that.

Comment (iamcarbon) [reply src/SixLabors.Core/MathF.cs]: Yeah, thought about this. Good to set the prescient now. Updated the PR to define a symbol.

**Comment by iamcarbon** [comment]
Ready for review.

**Comment by iamcarbon** [comment]
Ping @JimBobSquarePants

**Comment by JimBobSquarePants** [comment]
Sorry chum, was busy fixing up the Middleware. Will give this another review in the morning.

File: src/SixLabors.Core/SixLabors.Core.csproj [hunk 2]
```diff
@@ -34,24 +25,28 @@
         <CodeAnalysisRuleSet>..\SixLabors.ruleset</CodeAnalysisRuleSet>
     </PropertyGroup>
 
+    <PropertyGroup Condition=" $(TargetFramework.StartsWith('netcoreapp2')) ">
```
Comment (JimBobSquarePants) [inline src/SixLabors.Core/SixLabors.Core.csproj]: `$(TargetFramework.StartsWith('netcoreapp') && !TargetFramework.StartsWith('netcoreapp1'))`?


That should protect us for NET Core 3+

File: src/SixLabors.Core/SixLabors.Core.csproj [hunk 3]
```diff
@@ -34,24 +25,28 @@
...
     
+        <AdditionalFiles Include="..\..\stylecop.json" />
+    </ItemGroup>
+ 
+    <ItemGroup>
        <PackageReference Include="System.Runtime.CompilerServices.Unsafe" Version="4.5.1" />
     </ItemGroup>  
 
-    <ItemGroup Condition=" '$(TargetFramework)' != 'netcoreapp2.1'">
+    <ItemGroup Condition=" '$(TargetFramework)' != 'netcoreapp2.1' ">
```
Comment (JimBobSquarePants) [inline src/SixLabors.Core/SixLabors.Core.csproj]: I guess this should use the same condition as above?

File: src/SixLabors.Core/Helpers/HashHelpers.cs [hunk 4]
```diff
@@ -24,5 +24,39 @@ public static int Combine(int h1, int h2)
...
+        /// </summary>
+        /// <param name="h1">The first </param>
+        /// <param name="h2">Hash code two</param>
+        /// <param name="h3">Hash code three</param>
+        /// <returns>Returns a hash code for the provided hash codes.</returns>
+        public static int Combine(int h1, int h2, int h3)
+        {
+            int hash = Combine(h1, h2);
+
+            hash = Combine(hash, h3);
```
Comment (JimBobSquarePants) [inline src/SixLabors.Core/Helpers/HashHelpers.cs]: Combine assignment with return.

File: src/SixLabors.Core/Helpers/HashHelpers.cs [hunk 5]
```diff
@@ -24,5 +24,39 @@ public static int Combine(int h1, int h2)
...
+        /// <param name="h4">Hash code four</param>
+        /// <returns>Returns a hash code for the provided hash codes.</returns>
+        public static int Combine(int h1, int h2, int h3, int h4)
+        {
+            int hash = Combine(h1, h2);
+
+            hash = Combine(hash, h3);
+            hash = Combine(hash, h4);
+
+            return hash;
```
Comment (JimBobSquarePants) [inline src/SixLabors.Core/Helpers/HashHelpers.cs]: Combine assignment with return.

File: src/SixLabors.Core/MathF.cs [hunk 6]
```diff
@@ -3,7 +3,7 @@
 
 using System.Runtime.CompilerServices;
 
-#if NETCOREAPP2_0
+#if SUPPORTS_MATHF
```
Comment (JimBobSquarePants) [inline src/SixLabors.Core/MathF.cs]: Better :+1:

File: src/SixLabors.Core/Memory/ArrayPoolMemoryAllocator.Buffer{T}.cs [hunk 7]
```diff
@@ -71,7 +71,7 @@ protected override void Dispose(bool disposing)
         /// <summary>
         /// The <see cref="IManagedByteBuffer"/> implementation of <see cref="ArrayPoolMemoryAllocator"/>.
         /// </summary>
-        private class ManagedByteBuffer : Buffer<byte>, IManagedByteBuffer
+        private sealed class ManagedByteBuffer : Buffer<byte>, IManagedByteBuffer
```
Comment (JimBobSquarePants) [inline src/SixLabors.Core/Memory/ArrayPoolMemoryAllocator.Buffer{T}.cs]: Good idea 👍

File: src/SixLabors.Core/Primitives/Point.cs [hunk 8]
```diff
@@ -258,16 +258,16 @@ public void Offset(int dx, int dy)
...
 
         /// <inheritdoc/>
-        public override int GetHashCode() => this.GetHashCode(this);
-
-        /// <inheritdoc/>
-        public override string ToString()
+        public override int GetHashCode()
         {
-            return $"Point [ X={this.X}, Y={this.Y} ]";
+            return HashHelpers.Combine(this.X.GetHashCode(), this.Y.GetHashCode());
```
Comment (JimBobSquarePants) [inline src/SixLabors.Core/Primitives/Point.cs]: Collapse expression on single line.

File: src/SixLabors.Core/Primitives/Rectangle.cs [hunk 9]
```diff
@@ -452,19 +448,10 @@ public override string ToString()
         }
 
         /// <inheritdoc/>
-        public override bool Equals(object obj) => obj is Rectangle && this.Equals((Rectangle)obj);
+        public override bool Equals(object obj) => obj is Rectangle other && this.Equals(other);
```
Comment (JimBobSquarePants) [inline src/SixLabors.Core/Primitives/Rectangle.cs]: Much better 👍

File: src/SixLabors.Core/Primitives/Rectangle.cs [hunk 10]
```diff
@@ -452,19 +448,10 @@ public override string ToString()
...
 
         /// <inheritdoc/>
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public bool Equals(Rectangle other) => this.X == other.X && this.Y == other.Y && this.Width == other.Width && this.Height == other.Height;
```
Comment (JimBobSquarePants) [inline src/SixLabors.Core/Primitives/Rectangle.cs]: One things I'd like to do with the library is really clean up operators. We do it wrong all over the place.

`==` should always call `IEquatable<T>.Equals(T other)` so this statement should be to remove the potential overhead and standardize. (It's probably inlined but I don't want to take the risk).

``` c#
public bool Equals(Rectangle other) => this.X.Equals(other.X) && this.Y.Equals(other.Y) && this.Width.Equals(other.Width) && this.Height.Equals(other.Height);
```

File: tests/SixLabors.Core.Tests/Helpers/HashHelpersTests.cs [hunk 11]
```diff
@@ -0,0 +1,28 @@
+﻿// Copyright (c) Six Labors and contributors.
+// Licensed under the Apache License, Version 2.0.
+
+using Xunit;
+
+namespace SixLabors.Tests.Helpers
```
Comment (JimBobSquarePants) [inline tests/SixLabors.Core.Tests/Helpers/HashHelpersTests.cs]: 👍

File: tests/SixLabors.Core.Tests/Primitives/PointFTests.cs [hunk 12]
```diff
@@ -5,12 +5,24 @@
...
 namespace SixLabors.Primitives.Tests
 {
     public class PointFTests
     {
+        [Fact]
+        public void CanReinterpretCastFromVector2()
+        {
+            var vector = new Vector2(1, 2);
+
+            PointF point = Unsafe.As<Vector2, PointF>(ref vector);
```
Comment (JimBobSquarePants) [inline tests/SixLabors.Core.Tests/Primitives/PointFTests.cs]: Great idea! 👍

File: src/SixLabors.Core/SixLabors.Core.csproj (hunk 2 above)
Comment (iamcarbon) [reply src/SixLabors.Core/SixLabors.Core.csproj]: Let's wait until we figure out what new targets we'll be adding to start utilizing intrinsics before we try to future proof this.  NETSTANDARD2.1 is also coming.

File: src/SixLabors.Core/Helpers/HashHelpers.cs (hunk 4 above)
Comment (iamcarbon) [reply src/SixLabors.Core/Helpers/HashHelpers.cs]: done

**Comment by iamcarbon** [comment]
@JimBobSquarePants -- updated. I should catch a few more nits in another pass.

File: src/SixLabors.Core/SixLabors.Core.csproj (hunk 2 above)
Comment (JimBobSquarePants) [reply src/SixLabors.Core/SixLabors.Core.csproj]: Yeah ok. Let's wait til we need to.

**Comment by JimBobSquarePants** [comment]
And it's in, thanks!!

**Pull request merged**